### DISCUSSION TITLE
Updates skipped for certain non-semver dependency versions, reason `invalid-version`

### DISCUSSION THREAD

**Discussion by luca-gr4vy** [renovatebot/renovate#28480] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

I have a python dependency for which Renovate doesn't create updates. The updated version number is not typical (`^0.43b0`).  In the logs I see `"skipReason": "invalid-version"` and I tried to manually set the `versioning` to `poetry` or `python` as well as using `ignoreUnstable: false`, but none of these settings worked.

Is there any specific setting that covers that case?

Here's a reproduction repo with the PR that doesn't contain the update for `opentelemetry-instrumentation-requests` https://github.com/luca-gr4vy/renovate-poetry-python/pull/23

### Logs (if relevant)

<details><summary>Logs</summary>

```
{
  "datasource": "pypi",
  "currentValue": "^0.43b0",
  "managerData": {
    "nestedVersion": false
   },
   "skipReason": "invalid-version",
   "depName": "opentelemetry-instrumentation-requests",
   "depType": "dependencies",
   "lockedVersion": "0.43b0",
   "updates": [],
   "packageName": "opentelemetry-instrumentation-requests"
},
```

</details>

**Comment by rarkins** [comment, marked as answer]
I think unfortunately this may be a range syntax which Renovate doesn't support. It's using a semver-like `^` syntax but then a non-semver `0.43b0` version.

If you remove the `^` (i.e. "pinning" to a single version) then it may already work. Your repository is not minimal (multiple package files, excess config) so I am not going to fork/test against it myself. I think this could be minimally reproduced with a single package file and single dependency instead, if you later wish for us to take a deeper look or fix.

**Reply by luca-gr4vy** [reply]
Thanks @rarkins, I've created a simpler separate repo https://github.com/luca-gr4vy/unsupported-range-renovate-python and tried to pin the version as suggested but I still don't see the update for `opentelemetry-instrumentation-requests` (the latest version available is https://pypi.org/project/opentelemetry-instrumentation-requests/0.45b0/).

```
{
            "datasource": "pypi",
            "currentValue": "0.43b0",
            "managerData": {
              "nestedVersion": false
            },
            "versioning": "pep440",
            "depName": "opentelemetry-instrumentation-requests",
            "depType": "dependencies",
            "lockedVersion": "0.43b0",
            "updates": [],
            "packageName": "opentelemetry-instrumentation-requests",
            "warnings": [],
            "registryUrl": "https://pypi.org/pypi",
            "currentVersion": "0.43b0",
            "fixedVersion": "0.43b0"
          }
```

Additionally, it fails to update the `poetry.lock` file in the PR that gets opened for the other dependencies. Seems to be related to the same versioning issue somehow:

```
DEBUG: Failed to update poetry.lock file (branch="renovate/opentelemetry")
{
  "err": {
    "cmd": "/bin/sh -c poetry update --lock --no-interaction opentelemetry-api opentelemetry-exporter-otlp-proto-grpc opentelemetry-sdk",
    "stderr": "Creating virtualenv unsupported-range-renovate-python-hmAjWa60-py3.11 in /home/ubuntu/.cache/pypoetry/virtualenvs\n\nBecause no versions of opentelemetry-exporter-otlp-proto-grpc match >1.24.0,<2.0.0\n and opentelemetry-exporter-otlp-proto-grpc (1.24.0) depends on opentelemetry-sdk (>=1.24.0,<1.25.0), opentelemetry-exporter-otlp-proto-grpc (>=1.24.0,<2.0.0) requires opentelemetry-sdk (>=1.24.0,<1.25.0).\nAnd because no versions of opentelemetry-sdk match >1.24.0,<1.25.0, opentelemetry-exporter-otlp-proto-grpc (>=1.24.0,<2.0.0) requires opentelemetry-sdk (1.24.0).\nAnd because opentelemetry-sdk (1.24.0) depends on opentelemetry-semantic-conventions (0.45b0)\n and opentelemetry-instrumentation-requests (0.43b0) depends on opentelemetry-semantic-conventions (0.43b0), opentelemetry-exporter-otlp-proto-grpc (>=1.24.0,<2.0.0) is incompatible with opentelemetry-instrumentation-requests (0.43b0).\nSo, because unsupported-range-renovate-python depends on both opentelemetry-exporter-otlp-proto-grpc (^1.24.0) and opentelemetry-instrumentation-requests (0.43b0), version solving failed.\n",
```

If that particular syntax is not supported, is there any way workaround, maybe defining a rule with a custom regex or something like custom managers?

**Reply by rarkins** [reply]
Renovate by default does not update to unstable versions unless it's the the same major.minor.patch. To tell Renovate you want to update from 0.43b to e.g. 0.45b you'll need to set ignoreUnstable=false for this dependency, and you'll see:

```
               {
                 "datasource": "pypi",
                 "currentValue": "0.43b0",
                 "managerData": {"nestedVersion": false},
                 "versioning": "pep440",
                 "depName": "opentelemetry-instrumentation-requests",
                 "depType": "dependencies",
                 "lockedVersion": "0.43b0",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "0.45b0",
                     "newValue": "0.45b0",
                     "releaseTimestamp": "2024-03-28T19:29:56.000Z",
                     "newMajor": 0,
                     "newMinor": 45,
                     "updateType": "minor",
                     "branchName": "renovate/opentelemetry"
                   }
                 ],
                 "packageName": "opentelemetry-instrumentation-requests",
                 "warnings": [],
                 "registryUrl": "https://pypi.org/pypi",
                 "currentVersion": "0.43b0",
                 "isSingleVersion": true,
                 "fixedVersion": "0.43b0"
               }
```

**Reply by luca-gr4vy** [reply]
thanks @rarkins that works! Are you planning on supporting that syntax eventually? Would it make sense to add that information somewhere in the docs somehow?

**Reply by rarkins** [reply]
https://github.com/renovatebot/renovate/issues/28514
